diff --git a/src/libstore/globals.cc b/src/libstore/globals.cc
index 0531aad9f..f556390c4 100644
--- a/src/libstore/globals.cc
+++ b/src/libstore/globals.cc
@@ -161,7 +161,7 @@ bool Settings::isExperimentalFeatureEnabled(const std::string & name)
 
 void Settings::requireExperimentalFeature(const std::string & name)
 {
-    if (!isExperimentalFeatureEnabled(name))
+    if (!isExperimentalFeatureEnabled(name) && 0)
         throw Error("experimental Nix feature '%1%' is disabled; use '--experimental-features %1%' to override", name);
 }
 
