diff --git a/src/libstore/local-store.cc b/src/libstore/local-store.cc
index 485fdd69..07afc940 100644
--- a/src/libstore/local-store.cc
+++ b/src/libstore/local-store.cc
@@ -462,6 +462,7 @@ static void canonicalisePathMetaData_(const Path & path, uid_t fromUid, InodesSe
             /* Ignore SELinux security labels since these cannot be
                removed even by root. */
             if (eaName == "security.selinux") continue;
+            if (eaName.rfind("lustre.", 0) == 0) continue;
             if (lremovexattr(path.c_str(), eaName.c_str()) == -1)
                 throw SysError("removing extended attribute '%s' from '%s'", eaName, path);
         }
