diff --git a/src/libstore/local-store.cc b/src/libstore/local-store.cc
index f45af2bac..74ed13094 100644
--- a/src/libstore/local-store.cc
+++ b/src/libstore/local-store.cc
@@ -191,7 +191,7 @@ LocalStore::LocalStore(const Params & params)
         {
             AutoCloseFD fd = open(reservedPath.c_str(), O_WRONLY | O_CREAT | O_CLOEXEC, 0600);
             int res = -1;
-#if HAVE_POSIX_FALLOCATE
+#if 0
             res = posix_fallocate(fd.get(), 0, settings.reservedSize);
 #endif
             if (res == -1) {
diff --git a/src/libutil/archive.cc b/src/libutil/archive.cc
index ed0eb2fb5..73b33c848 100644
--- a/src/libutil/archive.cc
+++ b/src/libutil/archive.cc
@@ -328,7 +328,7 @@ struct RestoreSink : ParseSink
         if (!archiveSettings.preallocateContents)
             return;
 
-#if HAVE_POSIX_FALLOCATE
+#if 0
         if (len) {
             errno = posix_fallocate(fd.get(), 0, len);
             /* Note that EINVAL may indicate that the underlying
