diff --git a/src/libstore/local-store.cc b/src/libstore/local-store.cc
index 485fdd69..ae21f53c 100644
--- a/src/libstore/local-store.cc
+++ b/src/libstore/local-store.cc
@@ -325,7 +325,7 @@ void LocalStore::openDB(State & state, bool create)
 
     /* Set the SQLite journal mode.  WAL mode is fastest, so it's the
        default. */
-    string mode = settings.useSQLiteWAL ? "wal" : "truncate";
+    string mode = "truncate"; //settings.useSQLiteWAL ? "wal" : "truncate";
     string prevMode;
     {
         SQLiteStmt stmt;
