diff --git a/src/libstore/build/local-derivation-goal.cc b/src/libstore/build/local-derivation-goal.cc
index 2182f0bb4..dfe8f2c18 100644
--- a/src/libstore/build/local-derivation-goal.cc
+++ b/src/libstore/build/local-derivation-goal.cc
@@ -59,6 +59,14 @@
 
 namespace nix {
 
+#undef NIX_PREFIX
+#undef NIX_LIBEXEC_DIR
+static const std::string SELF{canonPath(readLink("/proc/self/exe"))};
+static const std::string NIX_PREFIX{canonPath(dirOf(dirOf(SELF)))};
+static const std::string NIX_LIBEXEC_DIR{NIX_PREFIX + "/libexec/nix"};
+#define NIX_PREFIX (NIX_PREFIX)
+#define NIX_LIBEXEC_DIR (NIX_LIBEXEC_DIR)
+
 void handleDiffHook(
     uid_t uid, uid_t gid,
     const Path & tryA, const Path & tryB,
@@ -139,8 +147,12 @@ void LocalDerivationGoal::killChild()
             ::kill(-pid, SIGKILL); /* ignore the result */
             buildUser->kill();
             pid.wait();
-        } else
+        } else {
+            /* since all our builds are salloc/srun jobs we just send
+               a term signal to avoid leaving dead jobs in the queue */
+            pid.setKillSignal(SIGTERM);
             pid.kill();
+        }
 
         assert(pid == -1);
     }
@@ -2022,7 +2034,26 @@ void LocalDerivationGoal::runChild()
 
         posix_spawn(NULL, builder.c_str(), NULL, &attrp, stringsToCharPtrs(args).data(), stringsToCharPtrs(envStrs).data());
 #else
-        execve(builder.c_str(), stringsToCharPtrs(args).data(), stringsToCharPtrs(envStrs).data());
+
+        // Configure for nix-user-chroot slurm submission building
+        args.pop_front();
+        args.push_front(builder);
+        args.push_front("@STOREROOT@/nix");
+        args.push_front(NIX_LIBEXEC_DIR + "/nix-user-chroot");
+        args.push_front("@SRUN@");
+        if(env.count("PPN"))
+          args.push_front(env["PPN"]);
+        else
+          args.push_front(std::to_string(settings.buildCores));
+        args.push_front("-c");
+        if(env.count("MEMORY"))
+          args.push_front("--mem=" + env["MEMORY"]);
+        if(env.count("WALLTIME"))
+          args.push_front("--time=" + env["WALLTIME"]);
+        args.push_front(env["name"]);
+        args.push_front("-J");
+        args.push_front("salloc");
+        execve("@SALLOC@", stringsToCharPtrs(args).data(), stringsToCharPtrs(envStrs).data());
 #endif
 
         throw SysError("executing '%1%'", drv->builder);
